Tutorials/WinForms/Classic/implement-crud-functionality-with-xpo-objects.md

```diff
@@ -0,0 +1,169 @@
+        if (CustomerID.HasValue)
+            CustomerBindingSource.DataSource = UnitOfWork.GetObjectByKey<Customer>(CustomerID.Value);
+        else CustomerBindingSource.DataSource = new Customer(UnitOfWork);
+    }
+    ```
+* Open the `CustomersListForm` designer, select the `CustomersGridView` component in the **Properties** window, and double-click the `RowClick` event in the **Events** section to create the event handler.
+* Add this code to the event handler:
+    ``` csharp
+    using DevExpress.XtraGrid.Views.Grid
+    // ...
+    private void CustomersGridView_RowClick(object sender, RowClickEventArgs e) {
+        if (e.Clicks == 2) {
+            e.Handled = true;
+            int customerID = (int)CustomersGridView.GetRowCellValue(e.RowHandle, colOid);
+            using (EditCustomerForm form = new EditCustomerForm(customerID)) {
+                form.ShowDialog(this);
+            }
+        }
+    }
+    ```
+* Press the **F5** key to run the application, double-click a `GridView` record, and see the result.
+* Close both windows and open the `EditCustomerForm` designer.
+* Select `LayoutControl` and add `SimpleButton` from Toolbox. Add `EmptySpaceItem` elements to adjust the button position and size.
+* Rename the `simpleButton1` control to **btnSave** and set its `Text` property to **&Save** (the ampersand sign assigns a mnemonic command to the button).
+* Double click the `btnSave` control to add the `Click` event handler (you can use the **Properties** window to add it).
+* Add this code to the `btnSave_Click` event handler:
+    ```csharp
+    private void btnSave_Click(object sender, EventArgs e) {
+        UnitOfWork.CommitChanges();
+        Close();
+    }
+    ```
+* The [UnitOfwork.CommitChanges](https://docs.devexpress.com/XPO/DevExpress.Xpo.UnitOfWork.CommitChanges) method saves changes to the database. To refresh the CustomersList Form, change code in the `CustomersGridView_RowClick` method as follows: 
+    ```csharp
+    private void Form1_Load(object sender, EventArgs e) {
+        Reload();
+    }
+    private void CustomersGridView_RowClick(object sender, RowClickEventArgs e) {
+        if (e.Clicks == 2) {
+            e.Handled = true;
+            int customerID = (int)CustomersGridView.GetRowCellValue(e.RowHandle, colOid);
+            using (EditCustomerForm form = new EditCustomerForm(customerID)) {
+                form.ShowDialog(this);
+                Reload();
+                CustomersGridView.FocusedRowHandle = CustomersGridView.LocateByValue("Oid", form.CustomerID.Value);
+            }
+        }
+    }
+    private void Reload() {
+        CustomersBindingSource.DataSource = new XPCollection<Customer>(new Session());
+    }
+    ```
+* Use the Visual Studio **Refactor** tool to rename the `Form1_Load` event handler to `CustomersListForm_Load`. To do this, put the cursor at the method name and click the **Edit > Refactor > Rename** menu item or use **Ctrl+R,Ctrl+R**.   
```

Comment:
"put the cursor at" Put 'at'? Why 'at'?

Reply:
I removed the word "position", because I found that a cursor definition is the "position indicator". As for the preposition, I often heard phrases like this: "the cursor is at the end of the line". That is why I used "at". However, actually it is necessary to put the cursor **before** the method name. I changed the preposition to "before".

8027bcfe801b5b035a8ad8b52281fc6c215ef75a